controller/contour/status.go

```diff
@@ -0,0 +1,164 @@
+			Status:  metav1.ConditionFalse,
+			Reason:  "ContourUnavailable",
+			Message: "Contour deployment does not exist.",
+		}
+	}
+
+	if ds == nil {
+		return metav1.Condition{
+			Type:    operatorv1alpha1.ContourAvailableConditionType,
+			Status:  metav1.ConditionFalse,
+			Reason:  "ContourUnavailable",
+			Message: "Envoy daemonset does not exist.",
+		}
+	}
+
+	dsAvailable := ds.Status.NumberAvailable > 0
+	for _, cond := range deployment.Status.Conditions {
+		if cond.Type != appsv1.DeploymentAvailable {
+			continue
+		}
+		switch {
+		case cond.Status == corev1.ConditionTrue:
+			if !dsAvailable {
+				return metav1.Condition{
+					Type:    operatorv1alpha1.ContourAvailableConditionType,
+					Status:  metav1.ConditionFalse,
+					Reason:  "ContourUnavailable",
+					Message: "Envoy daemonset does not have minimum availability.",
+				}
+			} else {
+				return metav1.Condition{
+					Type:    operatorv1alpha1.ContourAvailableConditionType,
+					Status:  metav1.ConditionTrue,
+					Reason:  "ContourAvailable",
+					Message: "Contour has minimum availability.",
+				}
+			}
+		case cond.Status == corev1.ConditionFalse:
+			if !dsAvailable {
+				return metav1.Condition{
+					Type:   operatorv1alpha1.ContourAvailableConditionType,
+					Status: metav1.ConditionFalse,
+					Reason: "ContourUnavailable",
+					Message: fmt.Sprintf("Envoy daemonset does not have minimum availability. Contour %s",
+						strings.ToLower(cond.Message)),
+				}
+			} else {
+				return metav1.Condition{
+					Type:    operatorv1alpha1.ContourAvailableConditionType,
+					Status:  metav1.ConditionFalse,
+					Reason:  "ContourUnavailable",
+					Message: fmt.Sprintf("Contour %s", strings.ToLower(cond.Message)),
+				}
+			}
+		case cond.Status == corev1.ConditionUnknown:
+			return metav1.Condition{
+				Type:    operatorv1alpha1.ContourAvailableConditionType,
+				Status:  metav1.ConditionUnknown,
+				Reason:  fmt.Sprintf("ContourUnknown: %s", cond.Message),
+				Message: fmt.Sprintf("Contour status unknown. %s", cond.Message),
+			}
+		}
+	}
+
+	return metav1.Condition{
+		Type:    operatorv1alpha1.ContourAvailableConditionType,
+		Status:  metav1.ConditionUnknown,
+		Reason:  "ContourUnknown",
+		Message: "Contour status unknown.",
+	}
+}
+
+// mergeConditions adds or updates matching conditions, and updates
+// the transition time if details of a condition have changed. Returns
+// the updated condition array.
+func mergeConditions(conditions []metav1.Condition, updates ...metav1.Condition) []metav1.Condition {
+	now := metav1.NewTime(clock.Now())
+	var additions []metav1.Condition
+	for i, update := range updates {
+		add := true
+		for j, cond := range conditions {
+			if cond.Type == update.Type {
+				add = false
+				if conditionChanged(cond, update) {
+					conditions[j].Status = update.Status
+					conditions[j].Reason = update.Reason
+					conditions[j].Message = update.Message
+					conditions[j].LastTransitionTime = now
```

Comment:
"lastTransitionTime is the last time the condition transitioned from one status to another."  So we should only update `conditions[j].LastTransitionTime` if `cond.Status != update.Status`.  (We have arguably been doing it incorrectly in other operators.)

https://github.com/kubernetes/apimachinery/blob/000b5f4f8623fa1a54269300e8987dfbf70033ed/pkg/apis/meta/v1/types.go#L1390-L1391